Cmake

jgoppert opened the pull request.

jgoppert committed:
Work on cmake build.

jgoppert committed:
Fixed dependency on genmsg gencpp.

jgoppert committed:
Added priority to modules for cmake.

jgoppert committed:
Merge branch 'master' of github.com:PX4/Firmware into cmake-nuttx

jgoppert committed:
Work on nuttx build. Hijacked makefile for cmake build.

jgoppert committed:
Removed test build scripts for cmake.

jgoppert committed:
Changed sts in vi format to 4.

jgoppert committed:
Added more flags to match nuttx build.

jgoppert committed:
Got nuttx compiling.

jgoppert committed:
Build flag cleanup.

jgoppert committed:
Default to posix sitl build.

jgoppert committed:
Changed test inclusion requiments for posix test.

jgoppert committed:
Renamed flags to be shorter on command line.

mcharleb committed:
Refactorted the cmake files to use a per OS file

The file cmake/${OS}/px4_target_impl.cmake implements a set of
macros that is called from CMakeLists.txt

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

jgoppert committed:
Merge pull request #17 from mcharleb/cmake-one-file-per-os

Refactorted the cmake files to use a per OS file

jgoppert committed:
Added copyright.

mcharleb committed:
Add qurt support to cmake build

Made additional fixes found by clang while enabling the qurt build.

Unused variables in perf_counter.c and main.cpp were fixed.

Use of poll vs px4_poll was fixed in px4_simple_app.c

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

jgoppert committed:
Merge pull request #18 from mcharleb/cmake-qurt-new

Add qurt support to cmake build

jgoppert committed:
Merge branch 'master' of github.com:PX4/Firmware into cmake-nuttx

jgoppert committed:
Updated test to account for poll fix.

jgoppert committed:
Added accelsim, simulator, filter, conversion modules to cmake build.

jgoppert committed:
cmake add: adcsim, airspeedsim, barosim, gpssim, gyrosim, tonealrmsim

jgoppert committed:
Merge branch 'master' of github.com:PX4/Firmware into cmake

jgoppert commented:
Ran into some issue with the new sensor modules, see #21.

mcharleb committed:
Added support for multiple config files per OS

New build configs are in:

cmake/${OS}/${OS}-${BOARD}-${TARGET_LABEL}

i.e. posix-sitl-simple

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
Added support for multiple config files per OS

New build configs are in:

cmake/${OS}/${OS}-${BOARD}-${TARGET_LABEL}

i.e. posix-sitl-simple

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb commented:
Sent new pull request to add separate config files per target OS: https://github.com/jgoppert/Firmware/pull/22

jgoppert committed:
Merge branch 'cmake-separate-configs' of https://github.com/mcharleb/Firmware into mcharleb-cmake-separate-configs

Conflicts:
	cmake/posix/px4_target_impl.cmake

jgoppert committed:
Merge branch 'mcharleb-cmake-separate-configs' into cmake

jgoppert committed:
Fixed posix-sitl build.

jgoppert committed:
Fixed nuttx build.

mcharleb committed:
Adding more CMakeLists.txt for posix build

src/systemcmds/topic_listener/CMakeLists.txt is
broken as it needs to generate a file

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
Added more cmake support for posix build

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
More cmake support for posix build

topic_listener is disabled for now as it needs a generated cpp file.

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
Added cmake support for qurt

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
Fixed CMakeLIsts.txt for topic_listener

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
cmake: add git version header file generation

Uses src/include/build_git_version.h.in as a template to generate
build_git_version.h in the build dir

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

mcharleb committed:
cmake: get the git version using execute_process

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

jgoppert committed:
Merge pull request #24 from mcharleb/cmake_new_multiconfig-2

Cmake new multiconfig 2

jgoppert committed:
Added tool to autogenerate CMakeLists.txt from module.mk.

jgoppert commented:
@mcharleb, in bb1d583 I added autogeneration of the CMakeLists.txt from the module.mk's. I think we should make sure the CMakeLists.txt are how we want them before we do the autogeneration.

So let's look over the cmake/cmake_lists.jinja file and agree upon what we want that to look like before I generate all of the missing files.

jgoppert committed:
Improved arg ahndling for cmake converter.

jgoppert committed:
Fixed cmake converter error message.

jgoppert committed:
Fixed nuttx building.

mcharleb commented:
@jgoppert Is the plan to do a one time generation of the CMakeLists.txt files and check them in, or to not have them and always autogenerate them? Autogeneration is a good idea to save work.... but I just did 46 files manually!

jgoppert commented:
@mcharleb The idea was to save us time in the manual conversion. In the long run I plan to delete the module.mk's.

jgoppert commented:
@mcharleb, @LorenzMeier I think we need to create a cmake coding standard, I was looking at this as a start:

https://techbase.kde.org/Policies/CMake_Coding_Style

I would also like to add:
* Use hard tabs instead of spaces (to be consistent with the source code, i'm open if you think we should use spaces)
* never use macros (right now it is hard to see what each of these macros are actually doing, using functions with explicit arguments will resolve this)
* never set global variables in an included cmake module
* never set add_library or add_executable within an included cmake module or cmake function (just to make it clear at the top level what we are doing)

mcharleb commented:
@jgoppert @LorenzMeier My knowledge of cmake is about 2 weeks old so I am simply doing things in a way that I know how. I agree with use of tabs. I was not able to get functions to work and set global variables from the functions so that is why I used macros. I'm fine if we can rework those as functions. The variable caching model of cmake is not very clear to me, nor is how to modify a variable passed to a function.

jgoppert commented:
I think add_custom_target and add_custom_command should  also be avoided in cmake modules, at global scope, maybe we could wrap these in a function like:

add_git_submodule(git_nuttx, repo), where add_git_submodule is a function and git_nuttx is the target of add_custom_target

Right now, we have something like add_git_submodule(nuttx, repo), and then are referencing git_nuttx, which it isn't clear where this is coming from.

jgoppert commented:
@mcharleb 

Here is an example of a function for list_prepend. Not that I'm using function_parsing. At the end, I'm setting the OUT variable, which is a one value argument in the PARENT_SCOPE. This is how you pass arguments out of a function.

```bash
function(list_prepend)
	#message(STATUS "${ARGN}")
	function_args_default()
	set(multiValueArgs LIST)
	set(oneValueArgs OUT PATH)
	set(requiredArgs LIST OUT PATH)
	function_parsing(list_prepend "${ARGN}")
	foreach(FILE ${LIST})
		list(APPEND ${OUT} ${PATH}/${FILE})
	endforeach()
	set(${OUT} ${${OUT}} PARENT_SCOPE)
endfunction(list_prepend)
```

jgoppert commented:
Here is how you would use this funtion:

```cmake
	list_prepend(
		OUT MSG_FILES_IN
		LIST ${MSG_FILES}
		PATH ${CMAKE_SOURCE_DIR}/msg/)
```

jgoppert commented:
@mcharleb , what do you think of not having add_library within each module, just define a set of variables in the module, and then call add library within the foreach loop?

jgoppert commented:
@mcharleb maybe prepending the module name to each variable is overkill since each variable is already scoped in that directory

mcharleb commented:
@jgoppert So each module would create a list, tupple or map (not sure of my options in CMake) and pass that out?

add_library(${module} STATIC EXCLUDE_FROM_ALL ${SRCS})

would have to do something like modify a global list of module names 

set(${MODULE_NAMES} ${${MODULE_NAMES}} PARENT_SCOPE)

and create a new global list variable with name ${module}_SRCS

set(${module}_SRCS  ${SRCS} PARENT_SCOPE)

Both of these violate your rule:

"never set global variables in an included cmake module"

jgoppert commented:
@mcharleb Good catch, there is actually a remedy for this: 

http://www.cmake.org/cmake/help/v3.0/command/get_directory_property.html

This is what I used to grab the module properties in the existing code.

In cmake think of each CMakeLists.txt as a scope, and then nested CMakeLists.txt also have nested scope. It is ok to set variables in a CMakeLists.txt because they are not global, they have scope, and then to grab them from above use explicit calsl to get_directory_property.

mcharleb commented:
@jgoppert OK, thanks. Now I understand your previous comment and agree that prepending the module name to each variable is overkill since each variable is already scoped in that directory.

jgoppert commented:
@mcharleb I think including these ideas will help make our code more readable, but how should we proceed with the current branch state?

mcharleb commented:
We can either:

1) have people manually write the CMakeLists.txt files and create a style checker to validate new files and any changes to the existing CMakeLists.txt module files.

2) Make people create a data file that a CMakeLists.txt generator can read and generate all of the module CMakeLists.txt (could become unwieldy to handle exceptions).

To get to the initial state of CMakeList.txt module files we can try to use your generator for module,mk files and redo all the existing module CMakeLists.txt files.

jgoppert commented:
OK, I'll generate once with the jinja template above with the module name removed form the local variables. I'll keep the old CMakeLists.txt's around as CMakeLists.txt.backup so we can use them as reference easily.

I think the style checker is a good idea, and we could maybe use a modified version of my converter code to parse the existing CMakeLists.txt and then diff against the generated version.

jgoppert committed:
Updated jinja template for cmake modules.

jgoppert committed:
Generated cmakelists.txt.

jgoppert commented:
@mcharleb  Embedding the module.mk and backing up the CMakeLitst.txt to CMakeLists.txt.backup seemed more cumbersome that it was worth given that we can just do the git diff, so I took that out.

jgoppert commented:
@mcharleb I hate how cmake does functions, too complicated, this function should simplify it for you.

```cmake
function(parse_function_args)
	cmake_parse_arguments("IN" "" "" "OPTIONS;ONE_VALUE;MULTI_VALUE;REQUIRED;ARGN" "${ARGN}")
	cmake_parse_arguments("OUT" "${IN_OPTIONS}" "${IN_ONE_VALUE}" "${IN_MULTI_VALUE}" "${IN_ARGN}")
	foreach(arg ${IN_OPTIONS} ${IN_ONE_VALUE} ${IN_MULTI_VALUE})
		set(${arg} ${OUT_${arg}} PARENT_SCOPE)
	endforeach()
endfunction()

function(testfunc)
	parse_function_args(
		ONE_VALUE NAME
		MULTI_VALUE LIST
		ARGN ${ARGN})
	message(STATUS "name: ${NAME}")
	message(STATUS "list: ${LIST}")
endfunction()

testfunc(NAME "ME"
	LIST a b c d)
```

mcharleb commented:
I'm looking at the newly generated CMakeLists.txt files. Would it be possible to create
```
set(extra_cxxflags)
```
vs
```
set(extra_cxxflags
        )
```
When extra_cxxflags is empty?
```
set(module systemcmds_config)
set(main config)
set(priority )
set(stack_size 4096)
set(max_optimization -Os)
set(extra_cxxflags
        )
set(extra_cflags
        )
set(srcs
        config.c
        )
set(includes
        )
```

mcharleb commented:
Looks possible:
https://github.com/etianen/moody-templates/issues/6

Jinja does it like this:
```
{% if foo -%}
```
Where the minus sign means "strip trailing / leading whitespace". More info here: http://jinja.pocoo.org/docs/templates/#whitespace-control

mcharleb commented:
Never mind.... I tried fiddling with it. Its likely easier to clean it up with a sed script.

jgoppert commented:
You could also use if blocks in jinja  and have a separate statement if it is empty.

mcharleb committed:
Removed extra space from template-created CMakeLists.txt

Also removed CMakeLists.txt from makefiles directory

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

jgoppert committed:
Merge pull request #25 from mcharleb/cmake-remove-spaces

Removed extra space from template-created CMakeLists.txt

mcharleb committed:
Fixed empty cmake set commands

Added python script to fix up files with empty set commands in the
form:

set(foo
	)

These become:

set(foo)

Signed-off-by: Mark Charlebois <charlebm@gmail.com>

jgoppert commented:
@mcharleb 
OK. I am trying something totally different here, let me know what you think:

The basic idea is to add every single CMakeLists.txt file, but only build the ones you want by specifying the module names in the module list.
https://github.com/jgoppert/Firmware/blob/cmake-2/CMakeLists.txt

This usese a cmake function to basically create a new add_library type command, but specific for px4. I setting the main function, stack etc. as target properties, this allows us to retrieve them later without the directory property hack.
https://github.com/jgoppert/Firmware/blob/cmake-2/src/modules/commander/CMakeLists.txt

I also improved the px4_utils and added docs:
https://github.com/jgoppert/Firmware/blob/cmake-2/cmake/px4_utils.cmake

mcharleb commented:
I don't like the following syntax for specifying the modules:

set(module_list
examples_px4_simple_app
modules_uORB
)

You can't tell dir names from names with underscores in them.

The more that I try to follow the requirement that module CMakeLists.txt
can't call add_library(), the more I think that is broken.

I am having to specify the full paths to all source files when I call
add_library(module ... $(srcs)) from the top CMakeLists.txt file. That
defeats the whole scoping behavior of the module directories.

-Mark

On Thu, Sep 3, 2015 at 10:47 PM, James Goppert <notifications@github.com>
wrote:

> @mcharleb <https://github.com/mcharleb>
> OK. I am trying something totally different here, let me know what you
> think:
>
> The basic idea is to add every single CMakeLists.txt file, but only build
> the ones you want by specifying the module names in the module list.
> https://github.com/jgoppert/Firmware/blob/cmake-2/CMakeLists.txt
>
> This usese a cmake function to basically create a new add_library type
> command, but specific for px4. I setting the main function, stack etc. as
> target properties, this allows us to retrieve them later without the
> directory property hack.
>
> https://github.com/jgoppert/Firmware/blob/cmake-2/src/modules/commander/CMakeLists.txt
>
> I also improved the px4_utils and added docs:
> https://github.com/jgoppert/Firmware/blob/cmake-2/cmake/px4_utils.cmake
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jgoppert/Firmware/pull/20#issuecomment-137654150>.
>

mcharleb commented:
@jgoppert I don't like the following syntax for specifying the modules

```
set(module_list
examples_px4_simple_app
modules_uORB
)
```

mcharleb commented:
px4_add_nuttx_firmware() doesn't belong in px4_utils

jgoppert commented:
Understandable, give me a day or two to run with this and then check back in. I think I have some ideas to address your concerns above. Agreed that not calling add library in the module directory is a bad idea. Maybe we had some miscommunication on that point.

jgoppert commented:
Platform specific dependencis for a module I think should be handled with if blocks in the module cmakelists.txt.

mcharleb commented:
Your design looks cleaner, but I am running out of time to spend on this and have been spending my evenings and early mornings on this. I am trying to help, but it seems everything I am spending my time on is just wasted effort.

jgoppert commented:
Just take a break for now. I think it is important we bounce this around until it's working well and i think we are close. I've benefited from your help and input even if we have both wasted too much time typing out cmake that we ended up nuking.

LorenzMeier commented:
James, please keep in mind that if we cannot close this with the resources we have we will have to give up on it and abandon it completely. Should we maybe set up a call today and review the current state and your goals with this?

jgoppert commented:
Yeah, I think that is a good idea.

mcharleb commented:
I'll be in the office in about 1/2 hour and I'll be there til about 4:30pm PT and then I'll be off for the long weekend.

jgoppert committed:
Merge pull request #26 from mcharleb/cmake-remove-spaces-2

Fixed empty cmake set commands

jgoppert commented:
@mcharleb I think I'm ready for you to take a look at this:
https://github.com/jgoppert/Firmware/blob/cmake-2/CMakeLists.txt

mcharleb commented:
Can you open an issue or make a pull request so I can provide comments?

Why are there "if nuttx" sections in the top level CMakeLIsts file?
Shouldn't then be encapsulated in px_nuttx_utils.cmake?

Why would the top level CMakeLists.txt not just include px4_nuttx_rules.mk
which defines a set of functions that each OS target must implement and
then the top level CMakeLists.txt just calls those functions and passes in
the appropriate top level variables to set?

We had that design working earlier.

I am concerned about having too much "if ${OS} ==" code in the top
CMakefile.txt. That would make it as friendly or less friendly than the
current build system based on GNU make.

-Mark

On Mon, Sep 7, 2015 at 2:08 PM, James Goppert <notifications@github.com>
wrote:

> @mcharleb <https://github.com/mcharleb> I think I'm ready for you to take
> a look at this:
> https://github.com/jgoppert/Firmware/blob/cmake-2/CMakeLists.txt
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jgoppert/Firmware/pull/20#issuecomment-138377300>.
>
